### PR TITLE
feat(typing): add type hints

### PR DESCRIPTION
**Pull request by rafalkrupinski**
Didn't use TypeAlias bc it's python>=3.10 so it would require typing_extensions (can be added if necessary).

quality_and_fitness_parsed declares `rtype: (float,int)` but the second value adds value of `q`, which may be float.

### CODE REVIEW

**Review by vytas7 — CHANGES_REQUESTED** [review]
Wow, we were not really planning much development here, but I agree that adding type hints would be a great addition! (We are in the process of adding type hints to the Falcon framework itself.) :+1:

Thank you for this improvement! :100:

However, shouldn't we include an empty `py.typed` file indicating this package is typed? Not even sure how to achieve that in the current module structure... maybe we need to create a subdirectory with `__init__.py` like other "normal" packages do? :thinking: 

Furthermore, running `mypy` tells me there is room for improvement:
```
$ mypy --strict mimeparse.py 
mimeparse.py:93: error: Argument 2 to "pop" of "dict" has incompatible type "None"; expected "str"  [arg-type]
mimeparse.py:134: error: Incompatible types in assignment (expression has type "int", variable has type "Literal[100, 0]")  [assignment]
mimeparse.py:141: error: Incompatible types in assignment (expression has type "int", variable has type "Literal[100, 0]")  [assignment]
mimeparse.py:144: error: Incompatible types in assignment (expression has type "float", variable has type "Literal[100, 0]")  [assignment]
mimeparse.py:148: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
```

**Comment by rafalkrupinski** [comment]
> Wow, we were not really planning much development here

I wanted to do #50 and it always helps to know what's being passed around.

> shouldn't we include an empty py.typed file

I'll take a look what to do about `py.typed` file in this case. Might need to move code to `mimeparse/__init__.py`, like you've said.

> Furthermore, running mypy tells me there is room for improvement

I did check with mypy, and the _return_ type hints are correct, only at some point inside functions the variables don't match them.

For example a weight is initialized with `0` and then at some point a float is added, or dict value is set to `None` and then overridden with some other value  🤷.

I wanted to keep the change minimal so I've decided to keep it. I'm happy to change it if you're planning to add `pre-commit` with mypy ;)

**Comment by vytas7** [comment]
Hm, I'm no expert at typing, but I think it would be good to check Mypy's output and see if we can fix the code inside the functions too. 
Otherwise, if we just care about the user experience for _using_ `mimeparse`, maybe we should simply a stubs instead?

What do you think @CaselIT?

**Comment by rafalkrupinski** [comment]
> Wow, we were not really planning much development here

Do you consider this package as internal to falconry? No problem, I can always fork.

**Comment by vytas7** [comment]
> > Wow, we were not really planning much development here
> 
> Do you consider this package as internal to falconry? No problem, I can always fork.

No, not internal at all. We are actually thinking even to reimplement parts of it an unvendor it from Falcon completely.

By writing that, I just meant that this package didn't attract many contribution in the 8 years since the last release :slightly_smiling_face:

**Comment by rafalkrupinski** [comment]
```sh
$ mypy mimeparse
Success: no issues found in 1 source file
```

**Review by vytas7 — CHANGES_REQUESTED** [review]
W00t, great that `mypy` passes now.

It seems that you have slightly altered the behaviour of the package's functions though; although I'm not opposed to change them if they are for good, but it would be better to track these improvement separately from typing.

**Comment by rafalkrupinski** [comment]
Yeah, I see that. I thought the changes wouldn't cause any issue, didn't think to run the tests :/

**Comment by rafalkrupinski** [comment]
I use it in my project, and I really like having less code to maintain, so...

**Review by vytas7 — APPROVED** [review]
Looks good now I think, but I'd let @CaselIT take a look at this before merging.

File: mimeparse/__init__.py [hunk 1]
```diff
@@ -4,14 +4,16 @@
 __license__ = 'MIT License'
 __credits__ = ''
 
+from typing import Dict, Generator, Iterable, Tuple, Union
```
Comment (CaselIT) [inline mimeparse/__init__.py:7]: let's import also `from __future__ import annotations`

**Review by CaselIT — APPROVED** [review]

File: mimeparse/__init__.py (hunk 1 above)
Comment (rafalkrupinski) [reply mimeparse/__init__.py:7]: what for? 🤔

Comment (CaselIT) [reply mimeparse/__init__.py:7]: to enable postponed evaluation of the annotations. see https://peps.python.org/pep-0563/

**Review by vytas7 — APPROVED** [review]
LGTM thanks!

**Pull request merged**